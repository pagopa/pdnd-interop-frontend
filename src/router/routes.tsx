import React from 'react'
import { createBrowserRouter } from 'react-router-dom'
import type { RouteObject } from 'react-router-dom'
import {
  PartyRegistryPage,
  ConsumerEServiceCatalogPage,
  ProviderEServiceListPage,
  ProviderAgreementsListPage,
  ConsumerAgreementsListPage,
  ConsumerPurposesListPage,
  ConsumerClientListPage,
  ConsumerClientM2MListPage,
  ConsumerEServiceDetailsPage,
  TOSPage,
  ProviderEServiceDetailsPage,
  ConsumerAgreementDetailsPage,
  ProviderAgreementDetailsPage,
  ProviderPurposeDetailsPage,
  OperatorDetailsPage,
  LogoutPage,
  KeyDetailsPage,
  ConsumerClientCreatePage,
  SecurityKeyGuidePage,
  ConsumerPurposeCreatePage,
  NotificationsPage,
  ConsumerAgreementCreatePage,
  NotFoundPage,
  ConsumerClientManagePage,
  ConsumerPurposeEditPage,
  ProviderEServiceCreatePage,
  ProviderPurposesListPage,
  ConsumerPurposeDetailsPage,
} from '@/pages'
import { LANGUAGES } from '@/config/constants'
import { getKeys } from '@/utils/array.utils'
import RoutesWrapper from './components/RoutesWrapper'
import type { LocalizedRoutes } from './router.types'
import Redirect from './components/Redirect'

// https://stackoverflow.com/a/70067918 waiting for "satisfies" operator in Typescript 4.9
const makeType = <T extends LocalizedRoutes>(o: T) => o

const EmptyComponent = () => <></>

export const routes = makeType({
  LOGOUT: {
    PATH: { it: 'logout', en: 'logout' },
    LABEL: { it: 'Logout', en: 'Logout' },
    COMPONENT: LogoutPage,
    PUBLIC: true,
    AUTH_LEVELS: ['admin', 'api', 'security'],
  },
  TOS: {
    PATH: { it: 'termini-di-servizio', en: 'terms-of-service' },
    LABEL: { it: 'Termini di servizio', en: 'Terms of service' },
    COMPONENT: TOSPage,
    PUBLIC: true,
    AUTH_LEVELS: ['admin', 'api', 'security'],
  },
  SECURITY_KEY_GUIDE: {
    PATH: { it: 'generazione-chiavi', en: 'generate-keys' },
    LABEL: { it: 'Come caricare le chiavi di sicurezza', en: 'How to upload public keys' },
    COMPONENT: SecurityKeyGuidePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api', 'security'],
  },
  NOTIFICATION: {
    PATH: { it: 'notifiche', en: 'notifications' },
    LABEL: { it: 'Notifiche', en: 'Notifications' },
    COMPONENT: NotificationsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api', 'security'],
  },
  PROVIDE_ESERVICE_CREATE: {
    PATH: { it: 'erogazione/e-service/crea', en: 'provider/e-service/create' },
    LABEL: { it: 'Crea e-service', en: 'Create e-service' },
    COMPONENT: ProviderEServiceCreatePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE_ESERVICE_EDIT: {
    PATH: {
      it: 'erogazione/e-service/:eserviceId/:descriptorId/modifica',
      en: 'provider/e-service/:eserviceId/:descriptorId/edit',
    },
    LABEL: { it: 'Modifica e-service', en: 'Edit e-service' },
    COMPONENT: ProviderEServiceCreatePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE_ESERVICE_MANAGE: {
    PATH: {
      it: 'erogazione/e-service/:eserviceId/:descriptorId',
      en: 'provider/e-service/:eserviceId/:descriptorId',
    },
    LABEL: { it: 'Visualizza e-service', en: 'View e-service' },
    COMPONENT: ProviderEServiceDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE_ESERVICE_LIST: {
    PATH: { it: 'erogazione/e-service', en: 'provider/e-service' },
    LABEL: { it: 'I tuoi e-service', en: 'Your e-services' },
    COMPONENT: ProviderEServiceListPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE_AGREEMENT_READ: {
    PATH: {
      it: 'erogazione/richieste/:agreementId',
      en: 'provider/agreements/:agreementId',
    },
    LABEL: { it: 'Gestisci richiesta', en: 'Manage request' },
    COMPONENT: ProviderAgreementDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE_AGREEMENT_LIST: {
    PATH: { it: 'erogazione/richieste', en: 'provider/agreements' },
    LABEL: { it: 'Richieste di fruizione', en: 'Requests for use' },
    COMPONENT: ProviderAgreementsListPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE_PURPOSE_LIST: {
    PATH: { it: 'erogazione/finalita', en: 'provider/purposes' },
    LABEL: { it: 'Finalità', en: 'Purposes' },
    COMPONENT: ProviderPurposesListPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE_PURPOSE_DETAILS: {
    PATH: { it: 'erogazione/finalita/:purposeId', en: 'provider/purpose/:purposeId' },
    LABEL: { it: 'Visualizza finalità', en: 'View purpose' },
    COMPONENT: ProviderPurposeDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  PROVIDE: {
    PATH: { it: 'erogazione', en: 'provider' },
    LABEL: { it: 'Erogazione', en: 'Provider' },
    REDIRECT: 'PROVIDE_ESERVICE_LIST',
    COMPONENT: EmptyComponent,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api'],
  },
  SUBSCRIBE_CATALOG_VIEW: {
    PATH: {
      it: 'fruizione/catalogo-e-service/:eserviceId/:descriptorId',
      en: 'subscriber/e-service-catalog/:eserviceId/:descriptorId',
    },
    LABEL: { it: 'Visualizza e-service', en: 'View e-service' },
    COMPONENT: ConsumerEServiceDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security', 'api'],
  },
  SUBSCRIBE_CATALOG_LIST: {
    PATH: { it: 'fruizione/catalogo-e-service', en: 'subscriber/e-service-catalog' },
    LABEL: { it: 'Catalogo e-service', en: 'E-service catalog' },
    COMPONENT: ConsumerEServiceCatalogPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security', 'api'],
  },
  SUBSCRIBE_PURPOSE_CREATE: {
    PATH: { it: 'fruizione/finalita/crea', en: 'subscriber/purpose/create' },
    LABEL: { it: 'Crea finalità', en: 'Create purpose' },
    COMPONENT: ConsumerPurposeCreatePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin'],
  },
  SUBSCRIBE_PURPOSE_EDIT: {
    PATH: {
      it: 'fruizione/finalita/:purposeId/modifica',
      en: 'subrscriber/purpose/:purposeId/edit',
    },
    LABEL: { it: 'Modifica finalità', en: 'Edit purpose' },
    COMPONENT: ConsumerPurposeEditPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin'],
  },
  SUBSCRIBE_PURPOSE_DETAILS: {
    PATH: { it: 'fruizione/finalita/:purposeId', en: 'subscriber/purpose/:purposeId' },
    LABEL: { it: 'Gestisci singola finalità', en: 'Manage purpose' },
    COMPONENT: ConsumerPurposeDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_PURPOSE_LIST: {
    PATH: { it: 'fruizione/finalita', en: 'subscriber/purpose' },
    LABEL: { it: 'Le tue finalità', en: 'Your purpose' },
    COMPONENT: ConsumerPurposesListPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_CLIENT_OPERATOR_EDIT: {
    PATH: {
      it: 'fruizione/client/:clientId/operatori/:operatorId',
      en: 'subscriber/client/:clientId/operators/:operatorId',
    },
    LABEL: {
      it: 'Gestisci operatore del client e-service',
      en: 'Manage e-service client operator',
    },
    COMPONENT: OperatorDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_CLIENT_KEY_EDIT: {
    PATH: {
      it: 'fruizione/client/:clientId/chiavi/:kid',
      en: 'subscriber/client/:clientId/keys/:kid',
    },
    LABEL: {
      it: 'Gestisci chiave pubblica del client e-service',
      en: 'Manage e-service client public key',
    },
    COMPONENT: KeyDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_CLIENT_CREATE: {
    PATH: { it: 'fruizione/client/crea', en: 'subscriber/client/create' },
    LABEL: { it: 'Crea client e-service', en: 'Create e-service client' },
    COMPONENT: ConsumerClientCreatePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin'],
  },
  SUBSCRIBE_CLIENT_EDIT: {
    PATH: { it: 'fruizione/client/:clientId', en: 'subscriber/client/:clientId' },
    LABEL: { it: 'Gestisci client e-service', en: 'Manage e-service client' },
    COMPONENT: ConsumerClientManagePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_CLIENT_LIST: {
    PATH: { it: 'fruizione/client', en: 'subscriber/client' },
    LABEL: { it: 'I tuoi client e-service', en: 'Your e-service clients' },
    COMPONENT: ConsumerClientListPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_AGREEMENT_READ: {
    PATH: {
      it: 'fruizione/richieste/:agreementId',
      en: 'subscriber/agreements/:agreementId',
    },
    LABEL: { it: 'Gestisci richiesta', en: 'Manage request' },
    COMPONENT: ConsumerAgreementDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_AGREEMENT_LIST: {
    PATH: { it: 'fruizione/richieste', en: 'subscriber/agreements' },
    LABEL: { it: 'Le tue richieste', en: 'Your requests' },
    COMPONENT: ConsumerAgreementsListPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_AGREEMENT_EDIT: {
    PATH: {
      it: 'fruizione/richieste/:agreementId/modifica',
      en: 'subscriber/agreements/:agreementId/edit',
    },
    LABEL: { it: 'Modifica richiesta', en: 'Edit request' },
    COMPONENT: ConsumerAgreementCreatePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin'],
  },
  SUBSCRIBE_INTEROP_M2M_CLIENT_OPERATOR_EDIT: {
    PATH: {
      it: 'fruizione/interop-m2m/:clientId/operatori/:operatorId',
      en: 'subscriber/interop-m2m/:clientId/operators/:operatorId',
    },
    LABEL: {
      it: 'Gestisci operatore del client api interop',
      en: 'Manage interop api client operator',
    },
    COMPONENT: OperatorDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_INTEROP_M2M_CLIENT_KEY_EDIT: {
    PATH: {
      it: 'fruizione/interop-m2m/:clientId/chiavi/:kid',
      en: 'subscriber/interop-m2m/:clientId/keys/:kid',
    },
    LABEL: {
      it: 'Gestisci chiave pubblica del client api interop',
      en: 'Manage interop api client public key',
    },
    COMPONENT: KeyDetailsPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_INTEROP_M2M_CLIENT_CREATE: {
    PATH: { it: 'fruizione/interop-m2m/crea', en: 'subscriber/interop-m2m/create' },
    LABEL: { it: 'Crea client api interop', en: 'Create interop api client' },
    COMPONENT: ConsumerClientCreatePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin'],
  },
  SUBSCRIBE_INTEROP_M2M_CLIENT_EDIT: {
    PATH: { it: 'fruizione/interop-m2m/:clientId', en: 'subscriber/interop-m2m/:clientId' },
    LABEL: { it: 'Gestisci client api interop', en: 'Manage interop api client' },
    COMPONENT: ConsumerClientManagePage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE_INTEROP_M2M: {
    PATH: { it: 'fruizione/interop-m2m', en: 'subscriber/interop-m2m' },
    LABEL: { it: 'I tuoi client api interop', en: 'Your api interop clients' },
    COMPONENT: ConsumerClientM2MListPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  SUBSCRIBE: {
    PATH: { it: 'fruizione', en: 'subscriber' },
    LABEL: { it: 'Fruizione', en: 'Subscriber' },
    REDIRECT: 'SUBSCRIBE_CATALOG_LIST',
    COMPONENT: EmptyComponent,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'security'],
  },
  PARTY_REGISTRY: {
    PATH: { it: 'ente', en: 'party' },
    LABEL: { it: 'Anagrafica ente', en: 'Party registry' },
    COMPONENT: PartyRegistryPage,
    PUBLIC: false,
    AUTH_LEVELS: ['admin', 'api', 'security'],
  },
  NOT_FOUND: {
    PATH: { it: '404', en: '404' },
    LABEL: { it: 'Pagina non trovata', en: 'Page not found' },
    COMPONENT: NotFoundPage,
    PUBLIC: true,
    AUTH_LEVELS: ['admin', 'api', 'security'],
  },
} as const)

/**
 * Adapts the custom localized routes object to the react-router-dom RouteObject
 */
export function mapRoutesToReactRouterDomObject() {
  const reactRouterDOMRoutes: RouteObject[] = [
    {
      path: '/',
      element: <Redirect to="SUBSCRIBE_CATALOG_LIST" />,
      children: [
        ...getKeys(LANGUAGES).map((lang) => ({
          path: lang,
          element: <Redirect to="SUBSCRIBE_CATALOG_LIST" />,
        })),
      ],
    },
    {
      path: '/*',
      element: <Redirect to="NOT_FOUND" />,
    },
  ]

  getKeys(LANGUAGES).forEach((lang) => {
    const langRoutes = getKeys(routes).reduce((acc, next) => {
      const accCopy = [...acc]
      const route = routes[next]
      let Component = <route.COMPONENT />

      if ('REDIRECT' in route) {
        Component = <Redirect to={route.REDIRECT} />
      }

      accCopy.push({
        path: route.PATH[lang],
        element: Component,
      })

      return accCopy
    }, [] as Array<RouteObject>)

    reactRouterDOMRoutes.push({
      path: lang,
      element: <RoutesWrapper />,
      children: langRoutes,
    })
  })

  return reactRouterDOMRoutes
}

export const router = createBrowserRouter(mapRoutesToReactRouterDomObject(), { basename: '/ui' })
