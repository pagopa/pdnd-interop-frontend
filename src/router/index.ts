export { RouterProvider } from './components/RouterProvider'
export { useCurrentRoute } from './hooks/useCurrentRoute'

import * as _routes from './routes'

export const { useNavigate, useParams, useLocation, useAuthGuard, useGeneratePath } = _routes.hooks
export const { Link, Redirect, Breadcrumbs } = _routes.components
export const { getParentRoutes } = _routes.utils
export type RouteKey = _routes.RouteKey
