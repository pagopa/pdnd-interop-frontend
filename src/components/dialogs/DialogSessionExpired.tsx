import React from 'react'
import { Button, Dialog, DialogActions, DialogContent, DialogTitle } from '@mui/material'
import { useTranslation } from 'react-i18next'
import type { DialogSessionExpiredProps } from '@/types/dialog.types'
import { useNavigateRouter } from '@/router'

export const DialogSessionExpired: React.FC<DialogSessionExpiredProps> = () => {
  const ariaLabelId = React.useId()
  const ariaDescriptionId = React.useId()

  const { t } = useTranslation('shared-components', { keyPrefix: 'dialogSessionExpired' })
  const { navigate } = useNavigateRouter()

  const logout = React.useCallback(() => {
    navigate('LOGOUT')
  }, [navigate])

  React.useEffect(() => {
    const timer = setTimeout(logout, 2500)
    return () => clearTimeout(timer)
  }, [logout])

  return (
    <Dialog open aria-labelledby={ariaLabelId} aria-describedby={ariaDescriptionId} fullWidth>
      <DialogTitle id={ariaLabelId}>{t('title')}</DialogTitle>

      <DialogContent id={ariaDescriptionId}>{t('content.description')}</DialogContent>

      <DialogActions>
        <Button variant="contained" onClick={logout}>
          {t('actions.confirmLabel')}
        </Button>
      </DialogActions>
    </Dialog>
  )
}
