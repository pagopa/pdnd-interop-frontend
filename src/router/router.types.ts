import { routes } from '@/router/routes'
import { LangCode } from '@/types/common.types'
import { UserProductRole } from '@/types/party.types'
import { NavigateOptions } from 'react-router-dom'

export type RouteKey = keyof typeof routes
export type ExtractRouteParams<T> = string extends T
  ? Record<string, string>
  : T extends `${infer _Start}:${infer Param}/${infer Rest}`
  ? { [k in Param | keyof ExtractRouteParams<Rest>]: string }
  : T extends `${infer _Start}:${infer Param}`
  ? { [k in Param]: string }
  : undefined

export type RouteParams<T extends RouteKey> = ExtractRouteParams<typeof routes[T]['PATH']['it']>

type UrlParams = Record<string, string>

export type GetRouteUrl = <T extends RouteKey>(
  route: T,
  ...config: RouteParams<T> extends undefined
    ? [{ urlParams?: UrlParams }] | []
    : [{ params: RouteParams<T> } & { urlParams?: UrlParams }]
) => string

type ExtendedNavigateOptions = NavigateOptions & { urlParams?: UrlParams }

export type Navigate = <T extends RouteKey>(
  route: T,
  ...config: RouteParams<T> extends undefined
    ? [ExtendedNavigateOptions] | []
    : [{ params: RouteParams<T> } & ExtendedNavigateOptions]
) => void

export type RouteAuthLevel = 'any' | Readonly<Array<UserProductRole>>
export type RouteConfig = {
  PATH: Record<LangCode, Readonly<string>>
  LABEL: Record<LangCode, Readonly<string>>
  EXACT?: boolean
  COMPONENT: React.FC
  PUBLIC: boolean
  AUTH_LEVELS?: RouteAuthLevel
}

export type Routes = Record<string, RouteConfig>
