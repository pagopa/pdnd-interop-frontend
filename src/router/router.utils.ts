import { routes } from '@/router/routes'
import type { LangCode, ProviderOrConsumer } from '@/types/common.types'
import type { LocalizedRoutes, RouteKey } from '@/router/router.types'
import { generatePath, matchPath } from 'react-router-dom'
import { getKeys } from '@/utils/array.utils'
import memoize from 'lodash/memoize'
import identity from 'lodash/identity'
import sortBy from 'lodash/sortBy'
import { PUBLIC_URL } from '@/config/env'
import qs from 'qs'
import QueryString from 'qs'
import isEqual from 'lodash/isEqual'

/** Returns the localized path of the given routeKey and language  */
export function getLocalizedPath(routeKey: RouteKey, lang: LangCode) {
  return `${PUBLIC_URL}/${lang}/${routes[routeKey].PATH[lang]}`
}

/** Checks if the routeKey represent the given pathname */
function matchRouteKeyPath(pathname: string, lang: LangCode, routeKey: RouteKey) {
  return matchPath(getLocalizedPath(routeKey as RouteKey, lang), pathname)
}

/** Returns the routeKey of the given pathname */
export const getRouteKeyFromPath = memoize(function _getRouteKeyFromPath(
  pathname: string,
  lang: LangCode
) {
  let pathnameWithBaseUrl = pathname
  if (!pathnameWithBaseUrl.startsWith(PUBLIC_URL)) {
    pathnameWithBaseUrl = PUBLIC_URL + pathnameWithBaseUrl
  }
  const isTheRouteKeyPath = matchRouteKeyPath.bind(null, pathnameWithBaseUrl, lang)
  const routeKey = getKeys(routes).find(isTheRouteKeyPath)

  if (!routeKey) {
    throw new Error(`Pathname ${pathnameWithBaseUrl} has no associated routeKey.`)
  }

  return routeKey
})

export function switchPathLang(fromLang: LangCode, toLang: LangCode) {
  for (const routeKey of Object.keys(routes)) {
    const match = matchRouteKeyPath(window.location.pathname, fromLang, routeKey as RouteKey)

    if (match) {
      let newPath = generatePath(
        getLocalizedPath(routeKey as RouteKey, toLang),
        match.params as Record<string, string>
      )

      if (window.location.search) {
        newPath += window.location.search
      }

      if (window.location.hash) {
        newPath += window.location.hash
      }

      window.location.replace(newPath)
      return
    }
  }
}

export const URL_FRAGMENTS: Record<string, Record<LangCode, string>> = {
  FIRST_DRAFT: { it: 'prima-bozza', en: 'first-draft' },
  EDIT: { it: 'modifica', en: 'bozza' },
}

export function getPathSegments(path: string) {
  return path.split('/').filter(identity)
}

export const getParentRoutes = memoize((routeKey: RouteKey): Array<RouteKey> => {
  function isParentRoute(
    possibleParentRouteSubpaths: Array<string>,
    currentSubpaths: Array<string>
  ) {
    if (possibleParentRouteSubpaths.length >= currentSubpaths.length) {
      return false
    }

    const allSameFragments = possibleParentRouteSubpaths.every(
      (pathFragment, i) => pathFragment === currentSubpaths[i]
    )

    // URL_FRAGMENTS.EDIT is appended at the end of a read path of the same type.
    // E.g. /eservice/myid becomes /eservice/myid/edit. So it's always same length + 1
    const isFalseParent = currentSubpaths.length === possibleParentRouteSubpaths.length + 1

    if (allSameFragments && isEditPath(routeKey) && isFalseParent) {
      return false
    }

    return allSameFragments
  }

  const route = routes[routeKey]
  const currentSubpaths = getPathSegments(route.PATH.it)

  const parents = Object.entries(routes).filter(([_, possibleParentRoute]) =>
    isParentRoute(getPathSegments(possibleParentRoute.PATH.it), currentSubpaths)
  )

  const sortedParents = sortBy(parents, ([_, parent]) => getPathSegments(parent.PATH.it).length)

  return sortedParents.map(([routeKey]) => routeKey) as Array<RouteKey>
})

export const isProviderOrConsumerRoute = memoize(
  (routeKey: RouteKey): ProviderOrConsumer | null => {
    const excludeList = ['ui', 'it']
    const subroutes = getPathSegments(routes[routeKey].PATH.it).filter(
      (b) => !excludeList.includes(b)
    )
    const [mode] = subroutes

    if (mode === 'erogazione') {
      return 'provider'
    }

    if (mode === 'fruizione') {
      return 'consumer'
    }

    return null
  }
)

export const isEditPath = memoize((routeKey: RouteKey): boolean => {
  const subroutes = getPathSegments(routes[routeKey].PATH.it)

  const lastBit = subroutes[subroutes.length - 1]
  return Object.values(URL_FRAGMENTS.EDIT).some((f) => lastBit.endsWith(f))
})

export function parseSearch(search: string) {
  return qs.parse(search, { ignoreQueryPrefix: true })
}

export function stringifySearch(searchObj: Record<string, string> | QueryString.ParsedQs) {
  return qs.stringify(searchObj)
}

const _getDynamicSegmentsFromPath = memoize((path: string) => {
  return path
    .split('/')
    .filter(identity)
    .filter((subpath) => subpath.startsWith(':'))
    .map((param) => param.replace(':', ''))
})

/**
 * Returns an array with all the dynamic path names for a given RouteKey
 * @example
 * "/:foo/test/:bar" => ["foo", "bar"]
 */
export const getDynamicPathSegments = memoize((routeKey: RouteKey) => {
  return _getDynamicSegmentsFromPath(routes[routeKey].PATH.it)
})

/**
 * For each language path in a LocalizedRoute, the dynamic path segments must be equal.
 * This function does a runtime check and throws if this requirement is not met for any of the implemented LocalizedRoute.
 * Optionally accepts a LocalizedRoutes object as argument for testing purposes.
 *
 * @example Okkay!
 *
 * ```json
 * {
 *    "it": "/:foo/route-italiana/:bar",
 *    "en": "/:foo/english-route/:bar",
 * }
 * ```
 *
 * @example Not okkay!
 * ```json
 * {
 *    "it": "/:foo/route-italiana/:bar",
 *    "en": "/:baz/english-route/:foo",
 * }
 * ```
 *
 */
export const checkLocalizedPathsConsistency = (_routes: LocalizedRoutes = routes) => {
  getKeys(_routes).forEach((routeKey) => {
    const paths = Object.values(_routes[routeKey].PATH)
    const firstPathDynamicSegments = _getDynamicSegmentsFromPath(paths[0])
    const firstPathSegmentNumber = getPathSegments(paths[0])

    const areLocalizedPathsConsistent = paths.every(
      (path) =>
        isEqual(_getDynamicSegmentsFromPath(path), firstPathDynamicSegments) &&
        isEqual(getPathSegments(path).length, firstPathSegmentNumber.length)
    )

    if (!areLocalizedPathsConsistent)
      throw new Error(
        `All the dynamic path segments for all the localized path (in the PATH property) must be equal. Check the ${routeKey} paths.`
      )
  })
}
