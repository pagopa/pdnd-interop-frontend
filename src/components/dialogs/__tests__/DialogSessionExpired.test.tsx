import React from 'react'
import { DialogSessionExpired } from '../DialogSessionExpired'
import { render } from '@testing-library/react'
import { vi } from 'vitest'
import * as router from '@/router'
import userEvent from '@testing-library/user-event'

const mockFn = vi.fn()

vi.spyOn(router, 'useNavigateRouter').mockReturnValue({
  navigate: mockFn,
  getRouteUrl: mockFn,
})

afterEach(() => {
  mockFn.mockClear()
})

describe('DialogSessionExpired testing', () => {
  it('should match the snapshot', () => {
    const screen = render(<DialogSessionExpired type="sessionExpired" />)
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should call navigate when click button', async () => {
    const screen = render(<DialogSessionExpired type="sessionExpired" />)
    const button = screen.getByRole('button', { name: 'actions.exitLabel' })
    const user = userEvent.setup()
    await user.click(button)
    expect(mockFn).toHaveBeenCalledWith('LOGOUT')
  })

  it('should call navigate after timeout', async () => {
    vi.useFakeTimers()
    render(<DialogSessionExpired type="sessionExpired" />)
    vi.advanceTimersByTime(2500)
    expect(mockFn).toHaveBeenCalledWith('LOGOUT')
  })
})
