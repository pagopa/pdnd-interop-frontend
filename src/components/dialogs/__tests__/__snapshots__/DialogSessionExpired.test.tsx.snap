// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DialogSessionExpired testing > should match the snapshot 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-describedby=":r1:"
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-tlc64q-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id=":r0:"
        >
          title
        </h2>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
          id=":r1:"
        >
          content.description
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            actions.exitLabel
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
